### DISCUSSION TITLE
hub dead, no lights, no reaction, is there still a chance?

### DISCUSSION THREAD

**Discussion by ertanden** [gpdaniels/spike-prime#41] [Q&A] [answered]
Hi, just saw this repo and wanted to ask here since you guys seem to know how this hub works.

My son has a Mindstorms 51515 Robot Inventor Hub (same as Spike hub I guess?) and a few days ago he told me when he was updating the firmware from the official Mindstorms Inventor app it was suddenly interrupted and there was an error.

After that the hub was not able to boot. I tried with USB and the power light came up and I could use pybricks (https://dfu.pybricks.com/) to flash the original 51515 firmware. The page displayed hub is "Rebooting" and after that absolute dead hub.

No matter what sequence of recovery buttons I press, or how long I press, or if I connect the USB cable, remove the battery, insert back the battery, charge for a few hours... I tried all the troubleshooting.

There's no light at all. No signs of life.

Did anybody here had the same experience? Do you see a chance of recovery? Any particular thing I should try?

Thanks! Appreciate all the work!

**Comment by dlech** [comment]
If you remove the battery, replace the battery, press and hold the Bluetooth button and the connect the USB cable to the hub and to a computer while still holding the Bluetooth button, does the Bluetooth light start to change colors?

Just like in this video: https://github.com/pybricks/support/issues/728#issuecomment-1293660327

**Comment by ertanden** [comment]
@dlech thanks! Actually seems like I can still perform the DFU flash using the BT button and USB cable. So there's still some life in the hub. 

However, I tried several times and actually multiple firmware versions from this repo, and the hub goes dead after the flash of a backup firmware. It looks like it doesn't even boot. No lights etc (replaced the battery etc...) and it can't connect to Mindsphere app using USB.

However, the strange thing is I tried to install the PyBricks Beta firmware (https://beta.pybricks.com/) and the hub seems to work fine with that! I even tried a small program to control some motors or read sensors etc. It just works with PyBricks firmware, but no luck with official firmware.

Any guess on this?

**Reply by dlech** [reply]
This could be because the file was not for the same hub or corrupted or something like that. We are working on integrating the firmware restore into the Pybricks Code app so it is as easy as flashing the Pybricks firmware. But it will be a week or two before that is ready. You could also try https://spike.legoeducation.com/hubdowngrade/.

**Reply by ertanden** [reply]
I tried nearly all the firmware versions (both for spike and storm) from this repo multiple times. Also trying flashing using the great `pybricksdev` CLI tool. But didn't make a difference. So in the end I guess this can't be resolved by just restoring the firmware.

There was a little hope, because some of the restored firmware versions caused the hub to be recognized on USB (Spike Prime app seemed to offer an update, but it immediately failed at %0). 

Also tried `rshell` but this can't connect;

```
Connecting to /dev/ttyACM0 (buffer-size 512)...
Trying to connect to REPL 
```

I feel like the filesystem may be corrupt. And have no idea how to fix it.

**Reply by dlech** [reply]
You could contact LEGO customer support and see what they say about the failed update.

**Reply by SpudGunMan** [reply]
try a different Pc with new lego OEM software. 

did it get black listed by accident in a ignore file or usb serial etc..

**Comment by ertanden** [comment, marked as answer]
I finally managed to fix the hub! Thanks @dlech for your support. **Pybricks** has so much potential and all the tooling around it is amazing. Also thanks @gpdaniels for this repo, these backup images helped me a lot for this recovery.

Let me document in detail what happened, so if anybody goes through similar issues, maybe it helps.

As I have written before, no matter how many times I flashed original backup firmware images using DFU, the hub was never booting up and never able to connect to Spike or Robot Inventor apps. Also **rshell** wasn't connecting...

However, at some point I tried the following sequence (after flashing firmware **spike - v1.3.00.0000-e8c274a** from this repo);

1. unplug USB cable
2. remove the battery
3. put battery back
4. press and hold the **Left** button (is this a known/documented sequence at all? I think I saw it mentioned somewhere when researching this issues, however I'm not sure...)
5. then connect the USB cable attached to the PC
6. Immediately I saw **rhshell** responded, and I knew something was different for good

After that Spike app and Robot Inventor app started recognizing the hub. To be honest, I tried the firmware  **spike - v1.3.00.0000-e8c274a** before, but since I never tried the above sequence with the "Left button" I was never able to connect from any app.

So after this stage, I performed the update provided by Spike app and the hub was then able to properly boot and connect to Spike Hub. But, I tried to run a small program and it complained **"No Space Left on Hub"**!

Then I tried, to **reset the hub** from Spike. After reset, the hub asked once again for an update, however as a result nothing changed. Still had "No space left on hub".

Finally, I switched to Robot Inventor app (first flashed the firmware image recommended in https://dfu.pybricks.com/), then connected with Robot Inventor app (this time I didn't need the "Left" button sequence). Performed the provided update from Robot Inventor app. But once again, when I tried to run a simple program: **"No Space Left on Hub"**!!!

Final thing that solved it for good. Perform **Hub reset** from Robot Inventor app. It performed the reset, asked for an update and voila! Now the hub seems to be back to normal and able to run programs from Robot Inventor app.

The hub is originally a Robot Inventor hub, so I guess the reset from Robot Inventor app could properly cleanup the storage, but Spike was not able to...

Anyways, that's it. :relieved:

**Reply by laurensvalk** [reply]
> press and hold the Left button (is this a known/documented sequence at all? I think I saw it mentioned somewhere when researching this issues, however I'm not sure...)

Good to hear it works again. Here's some extra info that possibly explains why this happens:

Sometimes the official SPIKE/MINDSTORMS firmware will not boot if the file system (containing your programs) is corrupted. These button combinations will wipe the file system and thus fix the problem.

We don't use the file system at all in Pybricks (since we don't want people to lose their files), so that might explain why it booted without problems.

**Reply by SpudGunMan** [reply]
That explains where some programs went!  I have been scratching my head for a while on a robot that keeps getting flashed blank! Thanks for this detail!

**Reply by laurensvalk** [reply]
It's probably very similar to the Pyboard reset: https://docs.micropython.org/en/latest/pyboard/tutorial/reset.html#safe-mode-and-factory-reset

The main difference is which button combination (left, right, left+right) does what. Maybe you can figure it out and share it :)

**Reply by gpdaniels** [reply]
Happy to add the different reset sequences to the readme of this repo if you figure them out!

**Reply by laurensvalk** [reply]
I found the following in my notes (it has been a while), but it would be good if someone wants to confirm:

- Hold left during boot: Do not mount file system, so skip executing `boot.py` and `main.py`.
- Hold right during boot: Mount file system, run `boot.py` but skip `main.py`.
- Hold left + right: Mount, but then reset the file system, creating a new `boot.py` and `main.py`.

I don't know if left+right will work if the file system is already corrupted. If it's the same as the Pyboard reset then it should probably work.

And if not, you should still be able to hold only left, and then use the REPL to reset the filesystem instead.

**Reply by ertanden** [reply]
Thanks @laurensvalk. This is extremely useful information and it makes sense that it worked for me with these sequences. So I can confirm somehow the "hold left during boot" sequence was the one that helped my case with a corrupt file system. The other ones though I can't try now, because my son wouldn't allow me... :D

**Reply by Ic33s** [reply]
Hi, have the same problem. => Half dead 51515 hub. 2 days ago I had a look at the referenced https://dfu.pybricks.com/ page and checked it out and set some bookmarks. Today I wanted to give it a shot, but: Unfortunatly today the information how to do the firmware update manually on windows/linux seems to be gone. There must happened some website update because some of my bookmark links to subpages are no longer working. Has someone saved the informations on how to do it on windows or linux? Thanks a lot.

**Reply by dlech** [reply]
Did you try the firmware restore on https://code.pybricks.com? Or are you looking for a command line tool?

**Reply by Ic33s** [reply]
Thanks for the quick reply. Finally I figured out by accident how to use the firmware restore. Completely my fault: I already had the hub booted up to DFU mode, because I still had the previous manual solution in mind and then there was no hub in the connection popup. Thank you very much. My son is very, very happy :)
